### PR TITLE
Chore: Add static analysis workflow

### PR DESCRIPTION
**Pull request by kjohnson**
## Description

This PR introduces static analysis to the automation pipeline.

`Give\DonorDashboards\Repositories\Donations`

- Adds a `function_exists` check for `give_pdf_receipts` in `getPaymentInfo` method.
The `class_exists` check for `Give_PDF_Receipts` is insufficient for static analysis.
- Adds a default `return` value to the `getIcon` method.

`Give\DonorDashboards\Tabs\EditProfileTab\AvatarRoute`

- Updates the `@return` type to include `WP_REST_Response` (in addition to `array`).
The WordPress REST API will convert an array to a `WP_REST_Response` automatically.

`Give\Donors\Endpoints\DeleteDonor`

- Casts `$ids` as integers to match intended type.

`Give\Exports\DonorsExport`

- Specifies parameter type for `give_export_get_data_{type}` filter in `filterExportData` method.

`Give\Form\LegacyConsumer\Commands\SetupFieldConfirmation`

- Adds `$payment`, `$receiptArgs` as class properties.

`Give\Form\LegacyConsumer\Commands\SetupFieldReceipt`

- Adds `$payment` as a class property.

`Give\Form\LegacyConsumer\FilterCallbackCollection`

- Adds `$callbacks` as a class property.

`Give\Form\Templates`

- Updates the parameter type to `array` for the `give_register_form_template` filter in the `load` method.
The parameter is a _keyed_ array, so typing the values does not work.

`Give\FormAPI\Fields`

- Adds missing parameter name in docblock for the `give_form_api_field_classes` filter in the `fromArray` method.

`Give\Framework\Database\DB`

- Adds default values to the `@method static update(...)` definition
Parameters 4 and 5 of `WPDB::update` are not required (with `null` default values) - this updates the `DB` wrapper to reflect that.

`Give\Framework\FieldsAPI`

- Removes unused `FieldsAPI/FormFieldMediator` code.
- Adds `ext-json` to `composer.json` for `JsonSerializable`.
- Updates constructors final to protect against  "[unsafe usage of new static()](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static)".

`Give\Framework\PaymentGateways`

- Updates abstract `PaymentCommand` constructor final to protect against  "[unsafe usage of new static()](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static)".
- Updates `DataTransferObjects/GatewayRouteData` class final to protect against  "[unsafe usage of new static()](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static)".

`Give\Log\Log`

- Adds a custom reflection extensions for the Log static facade. See `LogMethodsReflectionExtension`.

`Give\PaymentGateways\Gateways\PayPalStandard\Controllers\PayPalStandardWebhook`

- Updates order of parameters in docblock for filters in `supportLegacyActions` method.

`Give\PaymentGateways\Gateways\Stripe\Traits\SEPAMandateForm`

- Updates order of parameters in docblock for filters in `getMandateFormHTML` method.

`Give\PaymentGateways\Gateways\Stripe\CheckoutGateway`

- Specifies a default `case` for `switch` in the `getLegacyFormFieldMarkup` method.
While there are only two options, no default is provided, so this specifies "modal" as the default in conjunction with the method, `give_stripe_get_checkout_type`, called in the evaluated expression.

`Give\Receipt\Receipt`

- Adds the abstract method `validateSection`.
The method is assumed to be implemented by the extending class, but the contract is not defined in the base class.

`Give\ServiceProviders\LegacyServiceProvider`

- Adds default case in `is_request` method.
This might be better as an exception, but I don't want to introduce new code in this PR.

`Give\ServiceProviders\Onboarding`

- Removes container binding for undefined `DonationsRedirect` class.
Looking at the [relevant commit](f9f7988481da684e779b4cbcfd03d391b0599262), this binding was unnecessarily added. 

### Unsafe usage of new static

Many of the changes were related to the [unsafe usage of new static()](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static).

---

`Give\Framework\Models`

- Updates abstract `Model` constructor final to protect against  "[unsafe usage of new static()](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static)".

---

`Give\FormAPI\Form\Field`
`Give_Recurring_Cache` `LegacySubscriptions/includes/give-recurring-cache.php`
`Give\PaymentGateways\Stripe\Models\AccountDetail`

- Adds `final` constructor to protect against "[unsafe usage of new static()](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static)".

---

`Give\DonationForms\DataTransferObjects\DonationFormsResponseData`
`Give\Donations\DataTransferObjects\DonationQueryData`
`Give\Donations\DataTransferObjects\DonationResponseData`
`Give\Donations\Properties\BillingAddress`
`Give\Donors\DataTransferObjects\DonorQueryData`
`Give\Donors\DataTransferObjects\DonorResponseData`
`Give\Form\Template\Options`
`Give\FormAPI\Fields`
`Give\FormAPI\Section`
`Give\PaymentGateways\DataTransferObjects\FormData`
`Give\PaymentGateways\DataTransferObjects\GiveInsertPaymentData`
`Give\PaymentGateways\DataTransferObjects\SubscriptionData`
`Give\PaymentGateways\PayPalCommerce\Models\MerchantDetail`
`Give\PaymentGateways\Stripe\DataTransferObjects\DisconnectStripeAccountDto`
`Give\PaymentGateways\Stripe\DataTransferObjects\GetStripeAccountDetailsDto`
`Give\PaymentGateways\Stripe\DataTransferObjects\NewStripeAccountOnBoardingDto`
`Give\PaymentGateways\Stripe\DataTransferObjects\SetDefaultStripeAccountDto`
`Give\Subscriptions\DataTransferObjects\SubscriptionQueryData`
`Give\ValueObjects\Money (Deprecated)`

- Makes class `final` to protect against "[unsafe usage of new static()](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static)".
This class is not intended to be open for inheritance.

## Visuals

```neon
includes:
    - vendor/timeweb/phpstan-enum/extension.neon
    - vendor/szepeviktor/phpstan-wordpress/extension.neon

parameters:
    level: 0
    paths:
        - src/
    bootstrapFiles:
        - give.php
    scanDirectories:
        - includes/
    excludePaths:
        analyse:
            - src/LegacySubscriptions/includes/ # Scan but do not analyse
        analyseAndScan:
            - src/Form/LegacyConsumer/ServiceProvider.php # Needs a closer look
            - src/Framework/FieldsAPI/Concerns/HasType.php
            - src/Framework/Migrations/
            - src/Framework/PaymentGateways/
            - src/PaymentGateways/Actions/CreateSubscriptionAction/ # phpstan isn't finding inherited constructor for LegacySubscriber
            - src/TestData/ # Non-production code.
            - src/Tracking/Events/ThemeTracking.php # Unsure how undefined property $trackId is supposed to be declared or used.

services:
  -
    class: Give\PHPStan\Log\LogMethodsReflectionExtension
    tags:
      - phpstan.broker.methodsClassReflectionExtension

```

## Testing Instructions

The workflow runs on push to a branch

## Pre-review Checklist

-   [ ] ~Acceptance criteria satisfied and marked in related issue~
-   [ ] ~Relevant `@unreleased` tags included in DocBlocks~
-   [ ] I~ncludes unit tests~
-   [ ] ~Reviewed by the designer (if follows a design)~
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203201434282148

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@kjohnson Looking at how we're running the static analysis on GitHub, I see we have another workflow. I then wondered, "Why don't we just include this as a job in the tests workflow"? I then noticed that the tests workflow only runs on master and develop... but that doesn't seem right. What do you think? Is there a benefit of having this as a completely separate workflow? What if we just renamed the other one to Quality Assurance or something like that?

**Comment by kjohnson** [comment]
I think that I like the workflows being separate, but that is really just a preference.

They could certainly be in the same workflow.

- An additional job in the existing workflow would be outside of the test matrix, so no issue there.
- Jobs are able to run in parallel, so no problem there.

I could argue it both ways.

My preference is for separate, though, because they "change for different reasons", but even that is subjective,

**Comment by kjohnson** [comment]
> the tests workflow only runs on master and develop... but that doesn't seem right.

The tests run when a PR is opened against `develop`, so that covers almost everything. This means that we aren't triggering tests for every push to every branch, which can be a lot if you commit early and often, pushing those changes up.

**Comment by JasonTheAdams** [comment]
> The tests run when a PR is opened against `develop`, so that covers almost everything. This means that we aren't triggering tests for every push to every branch, which can be a lot if you commit early and often, pushing those changes up.

But it would run for every push to a PR going to one of those branches? I just don't want someone to know they're failing static analysis only after merging a PR.

**Comment by JasonTheAdams** [comment]
@kjohnson Also, we should probably have static analysis run on PHP 7.2 and PHP 8.0. I'd like to catch issues on our min and max PHP versions supported.

**Comment by kjohnson** [comment]
It would fail during the PR review process. It runs on pulls against `develop` and then again when merged into `develop` (which is a push on `develop`.

As an example, this un-merged PR shows all tests passing.

**Comment by kjohnson** [comment]
> we should probably have static analysis run on PHP 7.2 and PHP 8.0. I'd like to catch issues on our min and max PHP versions supported.

I can add a matrix to the new workflow.

**Comment by kjohnson** [comment]
> we should probably have static analysis run on PHP 7.2 and PHP 8.0. I'd like to catch issues on our min and max PHP versions supported.

- [ ] 7.2
- [ ] 7.4
- [ ] 8.0

Resolved in the shared workflow, see https://github.com/impress-org/givewp-github-actions/commit/da61eb193c02b4dcee3be419b6b6e3236adef573.

**Comment by kjohnson** [comment]
On a related note, there seems to be a lot of duplication between the "unit tests" workflow in the shared repo and the "wordpress tests" workflow in GiveWP. If we consolidated this to re-use the shared workflow then we could do something like this:

```yml
name: Static Analysis

on:
    push:
        branches: [ master, develop ]
    pull_request:
        branches: [ master, develop ]
    workflow_call: # Allows you to use this workflow as part of another workflow
    workflow_dispatch: # Allows you to run this workflow manually from the Actions tab

jobs:
    analyze:
        uses: impress-org/givewp-github-actions/.github/workflows/static-analysis.yml@master
    test:
        uses: impress-org/givewp-github-actions/.github/workflows/unit-tests.yml@master
```

File: src/PaymentGateways/Gateways/Stripe/CheckoutGateway.php
Comment (JasonTheAdams) [inline src/PaymentGateways/Gateways/Stripe/CheckoutGateway.php:119]: Is this intended to handle any other value? If not, default should throw an exception.

File: src/ServiceProviders/LegacyServiceProvider.php
Comment (JasonTheAdams) [inline src/ServiceProviders/LegacyServiceProvider.php:268]: Unless this is intended behavior, throw an exception.

File: phpstan.neon
Comment (JasonTheAdams) [inline phpstan.neon:20]: Why are we skipping theses?

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
Great work, @kjohnson! I left a few minor suggestions.

Regarding the test workflows, let's reconcile and clean up further DevOps stuff in following PRs.

File: .phpstan/Log/LogMethodsReflectionExtension.php
Comment (kjohnson) [reply .phpstan/Log/LogMethodsReflectionExtension.php]: Resolved by [a519724](https://github.com/impress-org/givewp/pull/6513/commits/a5197247add1aac8613ef3ba1c3ac53c03897692).

File: .phpstan/Log/LogStaticParameterReflection.php
Comment (kjohnson) [reply .phpstan/Log/LogStaticParameterReflection.php]: Resolved by [a519724](https://github.com/impress-org/givewp/pull/6513/commits/a5197247add1aac8613ef3ba1c3ac53c03897692).

File: .phpstan/Log/LogStaticMethodReflection.php
Comment (kjohnson) [reply .phpstan/Log/LogStaticMethodReflection.php]: Resolved by [a519724](https://github.com/impress-org/givewp/pull/6513/commits/a5197247add1aac8613ef3ba1c3ac53c03897692).

File: phpstan.neon
Comment (kjohnson) [reply phpstan.neon]: @JasonTheAdams these are ignored as the required changes are larger refactors than I wanted to include in this PR.

See [32132d6](https://github.com/impress-org/givewp/pull/6513/commits/32132d6ba5a852bbf61ec6f65cae536345616013).

File: src/ServiceProviders/LegacyServiceProvider.php
Comment (kjohnson) [reply src/ServiceProviders/LegacyServiceProvider.php]: On second look this is intended to fail silently to prevent an error, but such an error would be a result of developer error.

I pushed an update to enforce the supported request types, see [86fe7b8](https://github.com/impress-org/givewp/pull/6513/commits/86fe7b8102c0a305618bc6e5f7d79073c20cf38c).

File: src/PaymentGateways/Gateways/Stripe/CheckoutGateway.php
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/CheckoutGateway.php]: Just the two values, with `modal` being the default value.

I added an exception, see [c25fe0b](https://github.com/impress-org/givewp/pull/6513/commits/c25fe0b243f0febae3159b9b81ac0e7e0f4ef9a0).

File: phpstan.neon
Comment (JasonTheAdams) [reply phpstan.neon]: That's fair. Let's just make sure we chip away at these in future PRs until they're all done. 😃

**Review by JasonTheAdams — APPROVED** [review]
Great work, @kjohnson! I appreciate your sensibility to avoid major refactors in this PR. We'll continue to improve the scope of static analysis. I'm just glad we have it now and it will apply to all future work!

Please zip this up and send it off to QA for testing and approval.

**Comment by JasonTheAdams** [comment]
@kjohnson Where did we leave off with this? I'd like to get this bad boy merged!

**Comment by kjohnson** [comment]
@JasonTheAdams it looks like we left off with QA on August 29. I'm following up now.

Other than that I'll resolve the merge conflicts and have this ready for final approval.

See https://lw.slack.com/archives/C0355E6LBMZ/p1661789631719709

**Review by Benunc — APPROVED** [review]
This passes all automated tests, and manual inspection.

**Pull request merged**
